File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
TreeNode* ans=nullptr;               // Global Variable lelo 
void fun(TreeNode*root, TreeNode*p, TreeNode*q){
    // agar tree hai hi nhi yni root null hai 
    if(root==nullptr){
        return;
    }

    if(root==p || root==q){
        // agar root parent hi hai 
        ans=root;
        return;
    }

    if(root->val<p->val){
        // assume maan k chl rhe right se bhi chotta hai; yni parent root sabse chotta hai 
        fun(root->right,p,q);           // left wley pura cancel; right mei dekhlo apna LCA
    }
    else if(root->val>q->val){
        // yha par bhi root wla bnega nhi LCA
        fun(root->left,p,q);
    }
    else{
        // Agar root->val > p->val && root->val < q->val
        // matlab beech mei hai 
        ans=root;
        return;
    }
}
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {

        // chotti value ko p root bna do 
        if(p->val>q->val){
            swap(p,q);
        }
        
        // recursive Function call kardo 
        fun(root,p,q);

        // return the final ans
        return ans;
    }
};

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {

    TreeNode ans=null;               // Global Variable lelo

    public void fun(TreeNode root,TreeNode p,TreeNode q){
        // Agar Tree hai hi nhi yni root null hai
        if(root==null){
            return;
        }

        if(root==p || root==q){
            // Agar root parent hi hai
            ans=root;
            return;
        }

        if(root.val<p.val){
            // Assume maan k chl rhe right se bhi chotta hai; yni parent root sabse chotta hai
            fun(root.right,p,q);          // Left wla pura cancel; Right mei dekhlo apna LCA
        }
        else if(root.val>q.val){
            // Yha par bhi root wla bnega nhi LCA
            fun(root.left,p,q);
        }
        else{
            // Agar root.val > p.val && root.val < q.val
            // Matlab beech mei hai
            ans=root;
            return;
        }
    }

    public TreeNode lowestCommonAncestor(TreeNode root,TreeNode p,TreeNode q) {

        // Chhoti value ko p root bna do
        if(p.val>q.val){
            TreeNode temp=p;
            p=q;
            q=temp;
        }

        // Recursive Function Call kardo
        fun(root,p,q);

        // Return the Final Answer
        return ans;
    }
}